### PR TITLE
Processing of achievements through reflections

### PR DESCRIPTION
**Pull request by HarvelsX**
An asynchronous call to `nativeSyncPlayerAdvancements` works with Toasts, it is worth investigating this case.
Does not work for achievements related to statistics - planned to fix in the future.

### CODE REVIEW

**Comment by HarvelsX** [comment]
Oh, and it seems to remove the migrator :/

**Comment by WiIIiam278** [comment]
This is some good work, @HarvelsX. I'm somewhat keen on keeping the migrator in this plugin for now but can see the value of making an open source add-on that hooks into MPDB and this that does it itself to remove that compilation requirement.

I do wonder if it'd be worth packaging the NMS for each version in its own package and loading the class based on the parsed api version for better maintainability. I appreciate that this work doesn't remove the existing implementation as I'm quite strong on the opinion that the plugin shouldn't depend on NMS to work :)

I'd be happy to merge this if you're able to undo the migrator changes (as that work would be better for another PR down the line) and perhaps look into repackaging the NMS into packages (a la having the NMS package have a getter that returns the right class from sub packages; 1_17, 1_18 etc)

**Comment by HarvelsX** [comment]
> I do wonder if it'd be worth packaging the NMS for each version in its own package and loading the class based on the parsed api version for better maintainability

Hmm, this would be necessary for directed interactions with kernel classes, but in this case dynamic reflections are used. So it can be left that way. 
When an NMS class is initialized, it once gets elements of core classes and uses them in methods. If changed the names of core elements (classes, fields, methods), you can do a version check before retrieving items.

**Comment by HarvelsX** [comment]
> and perhaps look into repackaging the NMS into packages (a la having the NMS package have a getter that returns the right class from sub packages; 1_17, 1_18 etc)

You can, but then should switch to a directed interaction with the core. This would involve plugging in the `paperweight` plugin and creating an implementation for each version.
Do you think I should?

**Comment by WiIIiam278** [comment]
> When an NMS class is initialized, it once gets elements of core classes and uses them in methods. If changed the names of core elements (classes, fields, methods), you can do a version check before retrieving items.

Oh, I see! I think you actually already answered this question before, I forgot. My bad. In that case this implementation should be fine,  reflection is better than editing the core I think as it doesn't require build tools to compile the spigot dependency. In that case this does look quite good and paperweight shouldn't be necessary. My comment was more from a perspective of maintainability, in as so much that updating this to support future releases would only require version checks / modifications in one place, but as I understand from looking at this a bit closer you've done that 👍 

In that case, once you've done the tidy up with the migrator stuff and rebased ill review this for merge

**Comment by HarvelsX** [comment]
Now it works asynchronously and without Toasts

**Comment by HarvelsX** [comment]
Got MPDBMigrator back, you can consider this PR.

**Comment by HarvelsX** [comment]
Oh
```bash
[18:58:51 ERROR]: Entity threw exception at world:0.7224463201598326,71.0,0.06014237518302501
java.util.ConcurrentModificationException: null
        at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:756) ~[?:?]
        at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:778) ~[?:?]
        at net.minecraft.server.PlayerAdvancements.flushDirty(PlayerAdvancements.java:392) ~[?:?]
        at net.minecraft.server.level.ServerPlayer.tick(ServerPlayer.java:663) ~[?:?]
        at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1128) ~[?:?]
        at net.minecraft.world.level.Level.guardEntityTick(Level.java:970) ~[?:?]
        at net.minecraft.server.level.ServerLevel.lambda$tick$5(ServerLevel.java:670) ~[?:?]
        at net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:43) ~[paper-1
.18.1.jar:git-Paper-84]
        at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:650) ~[?:?]
        at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1607) ~[paper-1.18
.1.jar:git-Paper-84]
        at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:480) ~[p
aper-1.18.1.jar:git-Paper-84]
        at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1470) ~[paper-1.18.1
.jar:git-Paper-84]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1264) ~[paper-1.18.1.
jar:git-Paper-84]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.18
.1.jar:git-Paper-84]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

File: bukkit/src/main/java/me/william278/husksync/bukkit/config/ConfigLoader.java
Comment (WiIIiam278) [inline bukkit/src/main/java/me/william278/husksync/bukkit/config/ConfigLoader.java:29]: Use snake_case for config strings (use_native_implementation). Possibly also rename this option; `native_advancement_synchronization` or something?

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/nms/MinecraftVersionUtils.java
Comment (WiIIiam278) [inline bukkit/src/main/java/me/william278/husksync/bukkit/util/nms/MinecraftVersionUtils.java:16]: Just tidy up a bit of the styling here
`if (version == null) return 1;`
`for (int i = 0; i < length; i++) {...`

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java
Comment (WiIIiam278) [inline bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:166]: Change the formatting on this a bit too, try to stick to [google's code style](https://google.github.io/styleguide/javaguide.html).

```java
if (Settings.useNativeImplementation) {
    // ...
} else {
    // ...
}
```

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java
Comment (WiIIiam278) [inline bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:296]: If there's an error log with `plugin.getLogger().log(Level.SEVERE, "foo");`

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java
Comment (WiIIiam278) [inline bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:306]: Do you still plan on doing this?

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/nms/AdvancementUtils.java
Comment (WiIIiam278) [inline bukkit/src/main/java/me/william278/husksync/bukkit/util/nms/AdvancementUtils.java:68]: Code style; `public static Object getPlayerAdvancements(Player player) {`

**Review by WiIIiam278 — CHANGES_REQUESTED** [review]
Good stuff here, still a bit of stuff needing considering and cleaning up before merge + also fixing that async issue you found. I still need to take a bit of time to test thoroughly myself too. Also - can you undo the removal of the MPDB in this? Otherwise would require a manual merge.

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java
Comment (HarvelsX) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:166]: Please export the .editorConfig file, for code style

Comment (HarvelsX) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:166]: image

Comment (HarvelsX) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:166]: Export and publish to your repository for other developers

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java
Comment (HarvelsX) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:306]: Why not? I think this is what plugin users expect to see

Comment (WiIIiam278) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:306]: I was just wondering as there's still a TODO there :)

File: bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java
Comment (WiIIiam278) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:166]: Sure, can be done.

Comment (WiIIiam278) [reply bukkit/src/main/java/me/william278/husksync/bukkit/util/PlayerSetter.java:166]: https://drive.google.com/file/d/1_694N7UUw4tf9ixDSiKmnpp4tjXI4EPt/view?usp=sharing

**Review by WiIIiam278 — APPROVED** [review]
Yeah, looks pretty good now.
I gave it a test in game and it seemed to work pretty well, I think? It didn't show the toasts or anything (1.18). Did encounter an oddity that sometimes the advancements screen was stuck on "There is nothing to display here yet :(" even though it did seem to sync correctly but I reckon that's probably on my test environment.

**Comment by HarvelsX** [comment]
> Did encounter an oddity that sometimes the advancements screen was stuck on "There is nothing to display here yet :(" even though it did seem to sync correctly but I reckon that's probably on my test environment.
Yes, this happens in my production environment and I can not quite understand what's wrong, because on the test server everything is working fine :/ The next commit may fix it, can you help me and test it.

**Comment by WiIIiam278** [comment]
I've given it a bit of testing, it seems to work now, though I'm still not entirely sure what caused it in the first place.
This looks quite good to merge for me, provided you're happy. Provided it's ready, I'd like to merge soon.

**Comment by HarvelsX** [comment]
I couldn't figure out how to fix this problem, but it occurs when `markPlayerAdvancementsFirst` is called, i.e. when `IS_FIRST_PACKET` is set to `true`. I think you can merge this PR, but I'll still be looking for a solution to this problem for myself.

**Comment by HarvelsX** [comment]
If it doesn't work for you as well, just remove the asynchronous execution of `nativeSyncPlayerAdvancements` and also the `synchronized` structure in the data destination class.

**Review by WiIIiam278 — APPROVED** [review]
Excellent, this looks good to merge now. Just need to perform a few tests on my end and I'll get this merged before next update.

**Pull request merged**
